L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,259 @@
+        bool ICollection<KeyValuePair<string, object>>.Remove(KeyValuePair<string, object?> item);
+        void IDictionary<string, object>.Add(string key, object? value);
+        bool IDictionary<string, object>.ContainsKey(string key);
+        object? IDictionary<string, object>.get_Item(string key);
+        bool IDictionary<string, object>.Remove(string key);
+        void IDictionary<string, object>.set_Item(string key, object? value);
+        bool IDictionary<string, object>.TryGetValue(string key, out object? value);
+        IEnumerator<KeyValuePair<string, object?>> IEnumerable<KeyValuePair<string, object>>.GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator();
+    }
+    public readonly struct BalancerAttributesKey<TValue> {
+        public BalancerAttributesKey(string key);
+        public string Key { get; }
+    }
+    public class BalancerState {
+        public BalancerState(ConnectivityState connectivityState, SubChannelPicker picker);
+        public ConnectivityState ConnectivityState { get; }
+        public SubChannelPicker Picker { get; }
+    }
+    public class ChannelState {
+        public ChannelState(AddressResolverResult resolverState, BalancerAttributes options);
+        public BalancerAttributes Options { get; set; }
+        public AddressResolverResult ResolverState { get; set; }
+    }
+    public class CompleteContext {
+        public CompleteContext();
+        public Exception? Error { get; set; }
+        public Metadata? Trailers { get; set; }
+    }
+    public class DnsAddressResolver : AddressResolver {
+        public DnsAddressResolver(Uri address, ILoggerFactory loggerFactory);
+        public override Task RefreshAsync(CancellationToken cancellationToken);
+        public override IDisposable Subscribe(IObserver<AddressResolverResult> observer);
+        protected override void Dispose(bool disposing);
+    }
+    public class DnsAddressResolverFactory : AddressResolverFactory {
+        public DnsAddressResolverFactory(ILoggerFactory loggerFactory);
+        public override string Name { get; }
+        public override AddressResolver Create(Uri address, AddressResolverOptions options);
+    }
+    public abstract class LoadBalancer : IDisposable {
+        protected LoadBalancer();
+        public void Dispose();
+        public abstract void ResolverError(Exception exception);
+        public abstract void UpdateChannelState(ChannelState state);
+        public abstract void UpdateSubChannelState(SubChannel subChannel, SubChannelState state);
+        protected virtual void Dispose(bool disposing);
+    }
+    public abstract class LoadBalancerFactory {
+        protected LoadBalancerFactory();
+        public abstract string Name { get; }
+        public abstract LoadBalancer Create(IConnectionController controller, IDictionary<string, object> options);
+    }
+    public class PickContext {
+        public PickContext();
+        public HttpRequestMessage? Request { get; set; }
```

Comment:
Note that we don't actually have a complete HTTP request at this point; we may not have anything more than the headers to send.  In the case of a streaming RPC, there may not be any messages to send yet at this point.

Reply:
`HttpRequestMessage` isn't complete/sent at this point. It's the request message type the .NET client library.

As you said, it will just have headers at this point. Serialization of any messages will wait until it is sent.